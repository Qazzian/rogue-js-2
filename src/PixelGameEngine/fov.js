import Grid from './Grid';


export class Edge {
	constructor(x1, y1, dx, dy) {
		this.x1 = x1;
		this.x2 = x1;
		this.y1 = y1;
		this.y2 = y1;

		if (dx || dy) {
			this.extend(dx, dy);
		}
	}

	extend(dx, dy) {
		this.x2 += dx;
		this.y2 += dy;
	}

	getPoints() {
		return [
			[this.x1, this.x2],
			[this.y1, this.y2],
		];
	}
}

export function buildGeometry(mapTiles, isBlockingTest) {
	const parsedTiles = {};
	const edges = [];
	const grid = new Grid(mapTiles);

	try {
		grid.forEach((mapTile, x, y) => {
			const neighbours = {
				n: north(x, y),
				s: south(x, y),
				e: east(x, y),
				w: west(x, y),
			};

			if (isBlock(mapTile)) {
				const parsedBlock = {};
				if (isNotBlock(neighbours.n)) {
					if (isBlock(neighbours.w) && parsedTiles[`${x - 1}_${y}`].n) {
						parsedBlock.n = parsedTiles[`${x - 1}_${y}`].n;
						parsedBlock.n.extend(1, 0);
					} else {
						edges.push(new Edge(x, y, 1, 0));
						parsedBlock.n = edges[edges.length - 1];
					}
				}
				if (isNotBlock(neighbours.e)) {
					if (isBlock(neighbours.n && parsedTiles[`${x}_${y - 1}`].e)) {
						parsedBlock.e = parsedTiles[`${x}_${y - 1}`].e;
						parsedBlock.e.extend(0, 1);
					} else {
						edges.push(new Edge(x + 1, y, 0, 1));
						parsedBlock.e = edges[edges.length - 1];
					}
				}
				if (isNotBlock(neighbours.s)) {
					if (isBlock(neighbours.w && parsedTiles[`${x - 1}_${y}`].s)) {
						parsedBlock.s = parsedTiles[`${x - 1}_${y}`].s;
						parsedBlock.s.extend(1, 0);
					} else {
						edges.push(new Edge(x, y + 1, 1, 0));
						parsedBlock.s = edges[edges.length - 1];
					}
				}
				if (isNotBlock(neighbours.w)) {
					if (isBlock(neighbours.n && parsedTiles[`${x}_${y - 1}`].w)) {
						parsedBlock.w = parsedTiles[`${x}_${y - 1}`].w;
						parsedBlock.w.extend(0, 1);
					} else {
						edges.push(new Edge(x, y, 0, 1));
						parsedBlock.w = edges[edges.length - 1];
					}
				}
				parsedTiles[`${x}_${y}`] = parsedBlock;
			}
		});
	}
	catch (error) {
		console.error("error parsing geometry");
		console.info(JSON.stringify({grid, parsedTiles, edges}));
	}

	return edges;


	function isBlock(obj) {
		return obj !== undefined && isBlockingTest(obj);
	}
	function isNotBlock(obj) {
		return obj !== undefined && !isBlockingTest(obj);
	}

	function north(x, y) {
		return grid.get(x, y - 1);
	}

	function south(x, y) {
		return grid.get(x, y + 1);
	}

	function east(x, y) {
		return grid.get(x + 1, y);
	}

	function west(x, y) {
		return grid.get(x - 1, y);
	}
}
