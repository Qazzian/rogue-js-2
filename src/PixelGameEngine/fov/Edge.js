export default class Edge {
	constructor(x1, y1, dx, dy) {
		this.x1 = x1;
		this.x2 = x1;
		this.y1 = y1;
		this.y2 = y1;

		if (dx || dy) {
			this.extend(dx, dy);
		}
	}

	extend(dx, dy) {
		this.x2 += dx;
		this.y2 += dy;
	}

	getPoints() {
		return [
			[this.x1, this.x2],
			[this.y1, this.y2],
		];
	}
}
