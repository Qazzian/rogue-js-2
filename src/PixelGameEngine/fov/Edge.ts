export default class Edge {
	private readonly x1: number;
	private readonly y1: number;
	private x2: number;
	private y2: number;


	constructor(x1:number, y1:number, dx?:number, dy?:number) {
		this.x1 = x1;
		this.y1 = y1;

		this.x2 = x1;
		this.y2 = y1;

		if (typeof dx === 'number' && typeof dy === 'number') {
			this.extend(dx, dy);
		}
	}

	extend(dx:number, dy:number) {
		this.x2 += dx;
		this.y2 += dy;
	}

	getPoints() {
		return [
			{x: this.x1, y: this.y1},
			{x: this.x2, y: this.y2},
		];
	}
}
