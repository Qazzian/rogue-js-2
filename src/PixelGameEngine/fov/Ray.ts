export default class Ray {
  angle: number;
  dx: number;
  dy: number;

  constructor(angle: number, radius: number) {
    this.angle = angle;
    this.dx = radius * Math.cos(angle);
    this.dy = radius * Math.sin(angle);
  }
}
