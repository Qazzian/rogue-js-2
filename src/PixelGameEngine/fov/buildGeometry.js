import Grid from '../Grid';
import Edge from './Edge';


export default buildGeometry;

function buildGeometry(mapTiles, isBlockingTest) {
	const parsedTiles = {};
	const edges = [];
	const grid = new Grid(mapTiles);

	try {
		grid.forEach((mapTile, x, y) => {
			if (isBlock(mapTile)) {
				const neighbours = {
					n: north(x, y),
					s: south(x, y),
					e: east(x, y),
					w: west(x, y),
				};

				const parsedBlock = {};
				if (isNotBlock(neighbours.n)) {
					if (isBlock(neighbours.w) && parsedTiles[`${x - 1}_${y}`].n) {
						parsedBlock.n = parsedTiles[`${x - 1}_${y}`].n;
						parsedBlock.n.extend(1, 0);
					} else {
						edges.push(new Edge(x, y, 1, 0));
						parsedBlock.n = edges[edges.length - 1];
					}
				}
				if (isNotBlock(neighbours.e)) {
					if (isBlock(neighbours.n) && parsedTiles[`${x}_${y - 1}`].e) {
						parsedBlock.e = parsedTiles[`${x}_${y - 1}`].e;
						parsedBlock.e.extend(0, 1);
					} else {
						edges.push(new Edge(x + 1, y, 0, 1));
						parsedBlock.e = edges[edges.length - 1];
					}
				}
				if (isNotBlock(neighbours.s)) {
					if (isBlock(neighbours.w) && parsedTiles[`${x - 1}_${y}`].s) {
						parsedBlock.s = parsedTiles[`${x - 1}_${y}`].s;
						parsedBlock.s.extend(1, 0);
					} else {
						edges.push(new Edge(x, y + 1, 1, 0));
						parsedBlock.s = edges[edges.length - 1];
					}
				}
				if (isNotBlock(neighbours.w)) {
					if (isBlock(neighbours.n) && parsedTiles[`${x}_${y - 1}`].w) {
						parsedBlock.w = parsedTiles[`${x}_${y - 1}`].w;
						parsedBlock.w.extend(0, 1);
					} else {
						edges.push(new Edge(x, y, 0, 1));
						parsedBlock.w = edges[edges.length - 1];
					}
				}
				parsedTiles[`${x}_${y}`] = parsedBlock;
			}
		});
	}
	catch (error) {
		console.error("error parsing geometry");
		console.info(JSON.stringify(parsedTiles, edges));
	}

	return edges;


	function isBlock(obj) {
		return obj !== undefined && isBlockingTest(obj);
	}
	function isNotBlock(obj) {
		return obj !== undefined && !isBlockingTest(obj);
	}

	function north(x, y) {
		return grid.get(x, y - 1);
	}

	function south(x, y) {
		return grid.get(x, y + 1);
	}

	function east(x, y) {
		return grid.get(x + 1, y);
	}

	function west(x, y) {
		return grid.get(x - 1, y);
	}
}
