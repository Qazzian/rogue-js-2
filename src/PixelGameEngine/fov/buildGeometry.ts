import Grid from '../Grid';
import Edge from './Edge';

const DIRECTION = {
	n: [0, -1],
	e: [+1, 0],
	s: [0, +1],
	w: [-1, 0],
};

export default buildGeometry;

interface ProcessedBlock {
	n?: Edge | undefined,
	s?: Edge | undefined,
	e?: Edge | undefined,
	w?: Edge | undefined,
}


function buildGeometry(
	mapTiles: any[][],
	isBlockingTest: { (tile: any): any; (arg0: any): any; },
	addBorder: boolean
	): Edge[] {
	const parsedTiles:{[index: string]:any} = {};
	const edges: Edge[] = [];
	const grid = new Grid(mapTiles);

	if (addBorder) {
		edges.push(...addBoarders(grid));
	}

	try {
		grid.forEach((mapTile: any, x: number, y: number) => {
			if (isBlock(mapTile)) {
				const neighbours = {
					n: getNeighbour(x, y, DIRECTION.n),
					s: getNeighbour(x, y, DIRECTION.s),
					e: getNeighbour(x, y, DIRECTION.e),
					w: getNeighbour(x, y, DIRECTION.w),
				};

				const parsedBlock: ProcessedBlock = {};
				if (isNotBlock(neighbours.n)) {
					if (isBlock(neighbours.w) && parsedTiles[`${x - 1}_${y}`].n) {
						parsedBlock.n = parsedTiles[`${x - 1}_${y}`].n;
						parsedBlock.n!.extend(1, 0);
					} else {
						edges.push(new Edge(x, y, 1, 0));
						parsedBlock.n = edges[edges.length - 1];
					}
				}
				if (isNotBlock(neighbours.e)) {
					if (isBlock(neighbours.n) && parsedTiles[`${x}_${y - 1}`].e) {
						parsedBlock.e = parsedTiles[`${x}_${y - 1}`].e;
						parsedBlock.e!.extend(0, 1);
					} else {
						edges.push(new Edge(x + 1, y, 0, 1));
						parsedBlock.e = edges[edges.length - 1];
					}
				}
				if (isNotBlock(neighbours.s)) {
					if (isBlock(neighbours.w) && parsedTiles[`${x - 1}_${y}`].s) {
						parsedBlock.s = parsedTiles[`${x - 1}_${y}`].s;
						parsedBlock.s!.extend(1, 0);
					} else {
						edges.push(new Edge(x, y + 1, 1, 0));
						parsedBlock.s = edges[edges.length - 1];
					}
				}
				if (isNotBlock(neighbours.w)) {
					if (isBlock(neighbours.n) && parsedTiles[`${x}_${y - 1}`].w) {
						parsedBlock.w = parsedTiles[`${x}_${y - 1}`].w;
						parsedBlock.w!.extend(0, 1);
					} else {
						edges.push(new Edge(x, y, 0, 1));
						parsedBlock.w = edges[edges.length - 1];
					}
				}
				parsedTiles[`${x}_${y}`] = parsedBlock;
			}
		});
	}
	catch (error) {
		console.error("error parsing geometry");
		console.info(JSON.stringify({parsedTiles, edges}));
	}

	return edges;


	function isBlock(obj:any) {
		return obj !== undefined && isBlockingTest(obj);
	}
	function isNotBlock(obj: any) {
		return obj !== undefined && !isBlockingTest(obj);
	}

	function getNeighbour(x:number, y:number, direction:number[]) {
		return grid.get(x+direction[0], y+direction[1]);
	}
}

function addBoarders(grid: Grid) : Edge[] {
	const edges = [];
	edges.push(new Edge(0, 0, grid.width, 0));
	edges.push(new Edge(0, 0, 0, grid.height));
	edges.push(new Edge(grid.width, 0, 0, grid.height));
	edges.push(new Edge(0, grid.height, grid.width, 0));
	return edges;
}
