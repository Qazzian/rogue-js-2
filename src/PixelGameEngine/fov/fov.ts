import Edge from './Edge';

interface Point {
	x: number,
	y: number,
}

interface Ray {
	angle: number,
	x: number,
	y: number,
}

export default function fov(
	source: Point,
	geometry: Edge[],
	radius: number,
) {
	const lightRays = createRaysFromGeometry(source, geometry, radius);
	return calculateLightPolygon(lightRays);
}

export function createRaysFromGeometry(
	source: Point,
	geometry: Edge[],
	radius: number,) {
	return geometry.reduce((rays: Ray[], edge: Edge) => {
		const pointList = edge.getPoints();
		const rayList = pointList.reduce((allRays, point) => {
			return allRays.concat(createRaysFromPoint(source, radius, point));
		}, [] as Ray[]);

		return rays.concat(rayList);
	}, [] as Ray[]);
}

export function findLineIntersections(
	lightRays: Ray[],
	worldGeometry: Edge[]
) {
	lightRays.map((ray) => {
		const intersectedEdges = worldGeometry.filter((edge) => {
			const edgeVector = edge.getVector();
			if (doVectorsOverlap(edgeVector, ray)) {
				// TODO
			}
		})
	})
}

function doVectorsOverlap(v1: Point, v2: Point) {
	return Math.abs(v1.x - v2.x) === 0.0 || Math.abs(v1.y - v2.y) === 0.0
}

export function getIntersection(ray: Edge, segment:Edge){
// TODO translate into my interfaces
	// RAY in parametric: Point + Delta*T1
	const rayVector = ray.getVector();
	const segmentVector = segment.getVector();

	// Are they parallel? If so, no intersect
	const r_mag = Math.sqrt(
		rayVector.dx * rayVector.dx
		+ rayVector.dy * rayVector.dy
	);
	const s_mag = Math.sqrt(
		segmentVector.dx * segmentVector.dx
		+ segmentVector.dy * segmentVector.dy
	);
	if(rayVector.dx/r_mag == segmentVector.dx/s_mag
		&& rayVector.dy/r_mag == segmentVector.dy/s_mag){
		return null;
	}

	// SOLVE FOR T1 & T2
	// r_px+r_dx*T1 = s_px+s_dx*T2 && r_py+r_dy*T1 = s_py+s_dy*T2
	// ==> T1 = (s_px+s_dx*T2-r_px)/r_dx = (s_py+s_dy*T2-r_py)/r_dy
	// ==> s_px*r_dy + s_dx*T2*r_dy - r_px*r_dy = s_py*r_dx + s_dy*T2*r_dx - r_py*r_dx
	// ==> T2 = (r_dx*(s_py-r_py) + r_dy*(r_px-s_px))/(s_dx*r_dy - s_dy*r_dx)
	const T2 = (rayVector.dx*(segmentVector.y - rayVector.y)
		+ rayVector.dy*(rayVector.x - segmentVector.x))
		/(segmentVector.dx * rayVector.dy - segmentVector.dy*rayVector.dx);
	const T1 = (segmentVector.x + segmentVector.dx * T2 - rayVector.x)/rayVector.dx;

	// Must be within parametic whatevers for RAY/SEGMENT
	if(T1<0) return null;
	if(T2<0 || T2>1) return null;

	// Return the POINT OF INTERSECTION
	return {
		x: rayVector.x + rayVector.dx * T1,
		y: rayVector.y + rayVector.dy * T1,
		param: T1
	};

}

/**
 * Copied algorithm from http://www.ambrsoft.com/MathCalc/Line/TwoLinesIntersection/TwoLinesIntersection.htm
 * @param {Edge} ray
 * @param {Edge} segment
 */
export function findLineIntersection(ray: Edge, segment: Edge) {
	/**
	 * Line 1
	 * 	a = (y2 - y1);
			b = (x1 - x2);
			c = x1 * (y1 - y2) + y1 * (x2 - x1);
			m = (y2 - y1) / (x2 - x1);

	 		Line 2
			a_ = (y2_ - y1_);
			b_ = (x1_ - x2_);
			c_ = x1_ * (y1_ - y2_) + y1_ * (x2_ - x1_);
			m_ = (y2_ - y1_) / (x2_ - x1_);

			Lines Match
	 		((x1 == x2) && (y1 == y2)) || ((x1_ == x2_) && (y1_ == y2_))

	 		Lines Parallel
	 		(a * b_ == a_ * b)

	 		Intersection
			xi = (b * c_ - c * b_) / (a * b_ - b * a_);
			yi = (c * a_ - a * c_) / (a * b_ - b * a_);
	 */


}

export function createRaysFromPoint(source: Point, radius: number, destination: Point) : Ray[]  {
	const x = destination.x - source.x;
	const y = destination.y - source.y;
	const angle = Math.atan2(y, x);

	return [angle - 0.0001, angle, angle + 0.0001]
		.map((a) => makeRay(a, radius));
}

export function makeRay(angle: number, radius: number) {
	return {
		angle,
		x: radius * Math.cos(angle),
		y: radius * Math.sin(angle),
	};
}


