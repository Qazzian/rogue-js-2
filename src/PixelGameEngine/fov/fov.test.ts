import fov, {createRaysFromPoint, makeRay} from './fov';

import Edge from './Edge';

describe('FOV', () => {
	test('id defined', () => {
		expect(fov).toBeDefined();
		expect(typeof fov).toBe('function');
	});

	test('fov', () => {
		const geom = [new Edge(2, 2, 1, 0)];
		const rayList = fov({x: 1, y: 1}, geom, 5);
		expect(rayList).toBeDefined();
		expect(rayList.length).toBe(6);
		expect(rayList).toMatchSnapshot();
	});

	test('createRaysFromPoint', () => {
		const origin = {x: 0, y: 0};
		const testRays = createRaysFromPoint(origin, 10, {x: 1, y: 1});

		expect(testRays.length).toBe(3);
		expect(testRays).toMatchSnapshot();
	});


	test('makeRay', () => {
		expect(makeRay(0, 10))
			.toMatchObject({angle: 0, x: 10, y: 0});
		expect(makeRay(degToRad(360), 10))
			.toMatchObject({
				angle: 6.283185307179586,
				x: 10,
				y: -2.4492935982947065e-15,
			});
		expect(makeRay(degToRad(90), 10))
			.toMatchObject({
				angle: 1.5707963267948966,
				x: 6.123233995736766e-16,
				y: 10,
			});
	});
});


function degToRad(deg: number) {
	return deg * Math.PI / 180;
}
