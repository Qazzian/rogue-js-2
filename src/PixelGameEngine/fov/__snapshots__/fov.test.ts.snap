// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FOV createRaysFromPoint 1`] = `
Array [
  Object {
    "angle": 0.7852981633974483,
    "x": 7.071774883290144,
    "y": 7.070360669730128,
  },
  Object {
    "angle": 0.7853981633974483,
    "x": 7.0710678118654755,
    "y": 7.071067811865475,
  },
  Object {
    "angle": 0.7854981633974483,
    "x": 7.070360669730128,
    "y": 7.071774883290144,
  },
]
`;

exports[`FOV fov 1`] = `
Array [
  Object {
    "angle": 0.7852981633974483,
    "x": 3.535887441645072,
    "y": 3.535180334865064,
  },
  Object {
    "angle": 0.7853981633974483,
    "x": 3.5355339059327378,
    "y": 3.5355339059327373,
  },
  Object {
    "angle": 0.7854981633974483,
    "x": 3.535180334865064,
    "y": 3.535887441645072,
  },
  Object {
    "angle": 0.4635476090008061,
    "x": 4.472359539436277,
    "y": 2.235620752724695,
  },
  Object {
    "angle": 0.4636476090008061,
    "x": 4.47213595499958,
    "y": 2.23606797749979,
  },
  Object {
    "angle": 0.4637476090008061,
    "x": 4.471912325841522,
    "y": 2.2365151799142042,
  },
]
`;
